### ISSUE TITLE
Support blacklisting users

### ISSUE THREAD

**Issue by Half-Shot** [matrix-org/matrix-appservice-purple#33]
People aren't happy that @voyager:t2bot.io joins rooms over XMPP, so we should not connect & kick them from those rooms

**Comment by turt2live** [comment]
What are people not happy about? Voyager should categorically not be making people unhappy, and if there's something I can do from my end please let me know. 

Also, would blacklisting be bridge-wide, server-wide, or some other scope? If blacklisting is to be done, please do it aggressively such that the bot can take the hint in the future and better update the model used for projecting the size of Matrix.

**Comment by Half-Shot** [comment]
>What are people not happy about? Voyager should categorically not be making people unhappy, and if there's something I can do from my end please let me know.

Lots of people in the XMPP fundamentally object to having their rooms discovered by a bot and it would be best to ignore the XMPP rooms entirely. 

The blacklisting I'd propose would be kicking from a room when matching a filter in the config.

**Comment by Neustradamus** [comment]
Yes, it is important to be full XMPP compliance :)

**Comment by horazont** [comment]
>> What are people not happy about? Voyager should categorically not be making people unhappy, and if there's something I can do from my end please let me know.
> 
> Lots of people in the XMPP fundamentally object to having their rooms discovered by a bot and it would be best to ignore the XMPP rooms entirely.

That’s not true. *Discovering* rooms with a bot is fine (that’s what I do, see https://search.jabber.network, and I have yet to hear a complaint from a MUC operator or owner). That’s what public rooms are for.

We initially had the same idea as the voyager bot, apparently (there is even still code in the bot to actually enter rooms and watch them and track kicks and bans and whatnot). However, eventually we discarded it because it seemed too invasive for rather little gain. You can fairly easily discover rooms on a server, and operators can add their servers using existing in-protocol means (sending an invitation to the bot).

You also have to be careful to not enter/list rooms which aren’t meant to be publicly listed. I can imagine people being unhappy about *that* for sure. In XMPP, we have a flag for that (``muc_public``) which can be queried via disco#info (which is by the way independent of the flag for free-to-join-by-everyone). I don’t know whether bifrost transports this flag to Matrix and/or if there’s something equivalent in Matrix.

@Half-Shot Please be more careful with your wording, I’ve seen folks interpreting it as intentionally negative/FUDdy towards XMPP.
